hazelcast/src/main/java/com/hazelcast/util/Clock.java

```diff
@@ -71,15 +85,78 @@ static ClockImpl createClock() {
             return new SystemOffsetClock(offset);
         }
 
-        return new SystemClock();
+        if (XEN_CLOCK_ENABLED && XenOptimizedClock.xenClockDetected()) {
+            return new XenOptimizedClock();
+        } else {
+            return new SystemClock();
+        }
     }
 
+
     /**
      * Extend this class if you want to provide your own clock implementation.
      */
     public abstract static class ClockImpl {
 
         protected abstract long currentTimeMillis();
+
+        protected abstract long approximateTimeMillis();
+    }
+
+    /**
+     * Xen optimized clock.
+     */
+    public static class XenOptimizedClock extends ClockImpl {
+
+        private volatile long approximateTimeMillis = System.currentTimeMillis();
+
+        XenOptimizedClock() {
+            new ClockFetchThread().start();
+            System.out.println("Xen clocksource detected");
+        }
+
+        protected long currentTimeMillis() {
+            return System.currentTimeMillis();
+        }
+
+        protected long approximateTimeMillis() {
+            return approximateTimeMillis;
+        }
+
+        private class ClockFetchThread extends Thread {
+            ClockFetchThread() {
+                super("HZ-XenClockUpdater");
+                setDaemon(true);
+            }
+
+            @Override
+            public void run() {
+                while (true) {
+                    approximateTimeMillis = System.currentTimeMillis();
+                    try {
+                        Thread.sleep(XEN_CLOCK_UPDATE_PERIOD_MS);
+                    } catch (InterruptedException e) {
+                        EmptyStatement.ignore(e);
+                    }
+                }
+            }
+        }
+
+        private static boolean xenClockDetected() {
+            File file = new File("/sys/devices/system/clocksource/clocksource0/current_clocksource");
+            if (!file.exists()) {
+                System.out.println("/sys/devices/system/clocksource/clocksource0/current_clocksource does not exist");
+                return false;
+            }
+
+            try {
+                String clock = fileAsText(file);
+                return clock.toLowerCase().contains("xen");
+            } catch (Exception e) {
+                e.printStackTrace();
```

Comment:
please no printing on stdout/stderr

Reply:
In this case there is no logger available and I don't want to force loading a logger because the logger mechanism could get screwed up.